config/content.json

```diff
@@ -0,0 +1,49 @@
+{
```

Comment:
@jonpspri, is it possible to automatically generate content.json in the same fashion that build.sh (https://github.com/apache/incubator-openwhisk-cli/blob/master/build.sh) does?

Reply:
Yes, I think we can, but I'm not sure we want to.  The key configuration attributes are:
1.  OS Architecture combination,
2.  Filename for the tar file (with some interesting exception cases for naming standards).

I like having 'content.json' driving the build, because configuration directly drives content.  We _could_ add one layer of indirection so we don't manually maintain the directory structure and file names, but I'm not sure it's worth the work given that this configuration seems unlikely to change frequently.  Thoughts from any/all?

I will look at build.sh and also at what's over in incubator-openwhisk.  I welcome further input from you and others, especially on cases for where the Filename may need substitution.

Reply:
We can generate content.json from a few simple configuration arguments. For instance, build.gradle tells build.sh which OSes and architectures to build for. See https://github.com/apache/incubator-openwhisk-cli/blob/master/build.gradle#L57.

Reply:
The above process can be made more sophisticated in order to account for the additional linux architectures.

Reply:
Okay, but what's the future of the Gradle-based build?  From the README.md it looked like it was to be depricated in favor of a straight Go build.  And the Travis build for deployment runs the go builds directly in the Travis environment.  My thinking when I wrote this was that config.json was an already-defined container for the 'few simple configuration arguments' you mention above; however, if we're going to support gradle builds as well then yes, we can work out how to integrate the processes.|

I sincerely apologize if I'm coming off as argumentative -- that's not my intent.  I just want to make an informed decision about how to balance the various needs of the build (and consumption) processes.  So give me some time to digest what's going on with the Gradle/Docker build.

Reply:
@dubeejw Could you take a look at https://github.com/jonpspri/incubator-openwhisk/blob/2f90600617803cff7c4cae7d8880c1cad8e12160/ansible/roles/cli/files/createContentJson.sh ?  I essentially did there what would need to happen to build.sh in order to generate content.json .  In addtion, the arguments format for the docker build would need to change.  CLI_OS and CLI_ARCH as separate variables no longer works because once you introduce the more esoteric architectures like s390x and ppc64le, the build list is no longer a cartesian product of the two lists.  I've had to resort to lists of pairs instead to make the build work.  The reason I wasn't generating content.json is because it essentially IS that list of pairs.

Reply:
I'll have to take a closer look when I get a chacne. All input/conversation is welcome, argumentative or not. :)

Reply:
Meanwhile, as I've been poking around I've arrived at the conviction that the real problem is that we've got builds happening in two places:  'build.sh' (via gradle) and 'tools/travis/build_tag_releases.sh'.  So, I'm going to tinker with how the travis build works to see if I can get it to use 'build.sh'.  That will mean figuring out how to spec a cross-compile build via ./gradlew, though, which may bring us back around to the 'config.json' conversation.

Reply:
@houshengbo, can you provide any info regarding the above?

Reply:
I think we are fine with getting rid of the Docker build parts in favor of a straight Go build. The Docker build was previously used when the CLI resided in the main OpenWhisk repo. 

Perhaps we can use Gradle to kick off the Go build instead of Make.

Reply:
Okay, I'm going to start unwinding builds with a goal of having build logic reside in only one place and have something (Gradle or Make tbd) drive the build.  I'm up in the air about which build tool to use -- Gradle is great for the JVM-based languages with CLASSPATHs, etc, but Make may be a better answer for go, since it's more compiler-like.  We'll burn that bridge when we come to it -- first goal for me is eliminating the need for 'tools/travis/build_tag_releases.sh'.